Datahub.Sitemap/Datahub.Sitemap/SitemapGenerator.cs

```diff
@@ -0,0 +1,173 @@
+                BucketName = bucket,
+                Key = key,
+                InputStream = inputStream
+            });
+        }
+
+        /// <summary>
+        /// Generates an asset URL for a given asset, uses an optional BasePath and Scheme/Host from the Config object, outputs 
+        /// should look like;
+        /// 
+        /// Scheme(http|https)://Host/[Optional:BasePath]/:id
+        /// 
+        /// </summary>
+        /// <param name="id">The ID of the asset to create a URL for</param>
+        /// <param name="config">The Config Object for this lambda run</param>
+        /// <returns></returns>
+        private string GenerateAssetURL(string id, Parameters parameters)
+        {
+            var url = new UriBuilder
+            {
+                Host = parameters.Host,
+                Scheme = parameters.Scheme
+            }.Uri;
+
+            if (!string.IsNullOrWhiteSpace(parameters.BasePath))
+            {
+                url = url.Combine(parameters.BasePath);
+            }
+
+            return url.Combine(id).ToString();
+        }
+
+        /// <summary>
+        /// Create an XML element for a given asset (id/timestamp) combo, timestamp may be null so need to deal with it here
+        /// </summary>
+        /// <param name="xmlNS">The XML Namesapce to use</param>
+        /// <param name="parameters">Any input parameters to use</param>
+        /// <param name="id">The id of the asset</param>
+        /// <param name="timestamp">The timestamp of the asset</param>
+        /// <param name="changeFreq">(Optional) The change frequency of the asset (defaults to weekly)</param>
+        /// <returns></returns>
+        private XElement CreateElement(XNamespace xmlNS, Parameters parameters, string id, string timestamp, string changeFreq = "weekly")
+        {
+            return new XElement(xmlNS + "url",
+                    new XElement(xmlNS + "loc", GenerateAssetURL(id, parameters)),
+                    new XElement(xmlNS + "lastmod", timestamp),
+                    new XElement(xmlNS + "changefreq", changeFreq));
+        }
+
+        /// <summary>
+        /// Create the sitemap XML file from the scanned DynamoDB assets
+        /// </summary>
+        /// <param name="clientRequest">The ScanResponse from the DynamoDB table scan</param>
+        /// <param name="config">The Config Object for this lambda run</param>
+        /// <returns></returns>
+        public XDocument CreateSitemapXML(IEnumerable<Dictionary<string, string>> items, Parameters parameters)
+        {
+            XNamespace xmlNS = "http://www.sitemaps.org/schemas/sitemap/0.9";
+
+            return new XDocument(
+                new XDeclaration("1.0", "UTF-8", null),
+                new XElement(xmlNS + "urlset",
+                    from item in items
+                    select
+                    CreateElement(xmlNS, parameters, item.Single(x => x.Key == "id").Value, item.Single(x => x.Key == "timestamp").Value)
```

Comment:
`.Single` will throw an `InvalidOperationException` if there isn't exactly one result. That's 👍 but  if/when you get bad / unexpected data it can be hard to find the source of the prob. I'd move these two `.Single` calls on to their own lines so you get a line number in the exception message.

Reply:
Changed as suggested